feat: voice gateway v8 & xchacha20+aes256 encryption mode

Snazzah opened the pull request:
This is a WIP PR for using voice gateway V8 and using the "AEAD XChaCha20 Poly1305" encryption mode.

> [!CAUTION]
> This may deprecate tweetnacl, as that implementation does not have xchacha20.

- [x] `aead_xchacha20_poly1305_rtpsize`
  - [x] Audio sending
  - [x] Audio recieving
- [x] `aead_aes256_gcm_rtpsize` (using node:crypto)
  - [x] Audio sending
  - [x] Audio recieving
- [x] Voice Gateway V8
  - [x] Updated payloads
  - [x] Buffered resume

Snazzah committed:
feat: voice gateway v8 & xchacha20 encryption mode

TTtie commented on lib/voice/VoiceConnection.js:
```diff
@@ -22,7 +22,8 @@ try {
 let NaCl = null;
 
-const ENCRYPTION_MODE = "xsalsa20_poly1305";
+const VOICE_GATEWAY_VERSION = 8;
```
this might be worth moving into `lib/Constants.js` as `Constants.VOICE_VERSION` or something like that like is currently with REST and the regular gateway

Snazzah committed:
chore: move VOICE_VERSION to Constants

Snazzah committed:
fix: only recieve RTP packets with version 2

Snazzah committed:
fix: handle padded packets, fix nulled buffers

Snazzah committed:
fix: update post-decryption stuff with placeholder

Snazzah commented:
~~A bit of a problem, and a note to myself:
It seems like now after using the XChaCha20 mode, the decrypted packets when receiving seem to have changed how header extensions are given, it seems to follow [RFC5285 4.2](https://www.rfc-editor.org/rfc/rfc5285#section-4.2) without the first 4 bytes, so there isn't a good way of knowing how many one-byte headers there are.~~

The extension header is outside of the decrypted data because of `_rtpsize`, I get it now.

Snazzah committed:
fix: move nonce buffer to private class prop

Snazzah committed:
fix: finish voice recieve for xchacha20

Snazzah committed:
chore: simplify extension stripping

TTtie commented on lib/voice/VoiceConnection.js:
```diff
@@ -585,13 +596,36 @@ class VoiceConnection extends EventEmitter {
+            }
 
+            const hasExtension = !!(msg[0] & 0b10000);
+            const hasPadding = !!((msg[0] >> 5) & 1);
```
nit: keep this consistent with `hasExtension` and do the AND directly
```suggestion
            const hasPadding = !!(msg[0] & 0b100000);
```

Snazzah committed:
chore: code consistency

Co-authored-by: TTtie <me@tttie.cz>

Snazzah committed:
feat: add aes256_gcm mode

Snazzah committed:
fix: finish handling aes256_gcm recieving

TTtie commented on lib/voice/VoiceConnection.js:
```diff
@@ -163,6 +174,10 @@ class VoiceConnection extends EventEmitter {
     }
 
+    #aes256Available() {
+        return crypto && crypto.getCiphers().includes("aes-256-gcm");
```
it's 2024, use optional chaining
```suggestion
        return crypto?.getCiphers().includes("aes-256-gcm");
```

Snazzah committed:
chore: small style change

Co-authored-by: TTtie <me@tttie.cz>

Snazzah committed:
chore: comment fixes & clarifications

Snazzah committed:
feat: resume voice gateway handling

Snazzah committed:
refactor: remove tweetnacl

TTtie commented on lib/voice/VoiceConnection.js:
```diff
@@ -585,34 +639,61 @@ class VoiceConnection extends EventEmitter {
-                data = data.subarray(4 + l * 4);
+            if(hasExtension) {
+                const extOffset = headerSize - 4;
+                if(msg[extOffset + 0] == 0xbe && msg[extOffset + 1] == 0xde) { // RFC3550 5.3.1: RTP Header Extension
```
use strict equality (this isn't (and wasn't) enforced by linting yet)
```suggestion
                if(msg[extOffset + 0] === 0xbe && msg[extOffset + 1] === 0xde) { // RFC3550 5.3.1: RTP Header Extension
```

Snazzah committed:
fix: use strict equality

Co-authored-by: TTtie <me@tttie.cz>

Snazzah committed:
fix: properly handle reconnecting/resuming

voice server changes werent working properly

Snazzah committed:
fix: resume properly on ws close

TTtie commented on lib/voice/VoiceConnection.js:
```diff
@@ -189,11 +202,12 @@ class VoiceConnection extends EventEmitter {
             return;
         }
         this.channelID = data.channel_id;
+        const lastEndpoint = data.endpoint?.hostname;
```
typo? (`data.endpoint` is supposed to be [a string](https://discord.com/developers/docs/topics/gateway-events#voice-server-update-voice-server-update-event-fields) iirc)
```suggestion
        const lastEndpoint = this.endpoint?.hostname;
```

Snazzah committed:
fix: typo in connecting

Co-authored-by: TTtie <me@tttie.cz>

TTtie committed:
refactor: detect AES-256 when crypto gets loaded

querying for it just once should be enough I think

TTtie committed:
types: add VOICE_VERSION constant and VoiceConnection#resumed event

TTtie committed:
fix: use signed write for nonce

in JS, bitwise ops are signed and as such writeUInt32BE would cause an error on integer rollover

TTtie committed:
nvm, >>> 0 is more prevalent in the codebase, use that instead

TTtie commented on lib/voice/VoiceConnection.js:
```diff
@@ -820,26 +915,37 @@ class VoiceConnection extends EventEmitter {
+        } else if(Sodium) {
+            const ABYTES = Sodium.crypto_aead_xchacha20poly1305_ietf_ABYTES;
+            const length = frame.length + ABYTES;
+            this.sendBuffer.fill(0, headerSize, headerSize + ABYTES);
+            frame.copy(this.sendBuffer, headerSize + ABYTES);
+            Sodium.crypto_aead_xchacha20poly1305_ietf_encrypt(
+                this.sendBuffer.subarray(headerSize, headerSize + length),
+                this.sendBuffer.subarray(headerSize + ABYTES, headerSize + length),
```
i think this might be off (although it is working fine), the `ABYTES` should come after the encrypted data, not before it I think

TTtie committed:
zero the auth tag (ABYTES) at the correct place 

TODO find out if zeroing ABYTES is needed (will be probably defensively kept), but I think it probably isn't needed

TTtie committed:
nit: throw tweetnacl away from package.json browser object

TTtie reviewed (approved):
it all looks to be working, let's ship this tomorrow morning 🚀
